Replace secrets with join tokens

aaronlehmann opened the pull request:
Implement the proposal from
https://github.com/docker/docker/issues/24430#issuecomment-233100121

Removes acceptance policy and secret in favor of an automatically
generated join token that combines the secret, CA hash, and
manager/worker role into a single opaque string.

Adds a docker swarm join-token subcommand to inspect and rotate the
tokens.

Fixes #24430

cc @tonistiigi @diogomonica @tiborvass @aluzzardi

tonistiigi commented on api/client/swarm/opts.go:
```diff
@@ -15,29 +15,16 @@ import (
 type swarmOptions struct {
-	autoAccept          AutoAcceptOption
-	secret              string
+	token               string
```
What is this used for?

aaronlehmann replied:
It's the token that is passed to `swarm join -token ...`.

tonistiigi replied:
But join uses `joinOptions`?

aaronlehmann replied:
Missed that. I'll remove it from `swarmOptions`.

Maybe we should rename this to `initUpdateOptions`.

tonistiigi commented on daemon/cluster/cluster.go:
```diff
@@ -354,9 +351,6 @@ func (c *Cluster) Join(req types.JoinRequest) error {
 	for {
 		select {
 		case <-certificateRequested:
```
Should be safe to remove this whole case, certificateRequested var and for cycle.

tiborvass commented on api/client/swarm/join_token.go:
```diff
@@ -0,0 +1,71 @@
+	var rotate bool
+
+	cmd := &cobra.Command{
+		Use:   "join-token [--rotate] worker | manager",
```
i wonder if it should have parentheses `(worker|manager)` (like in docopt.org) not sure what's the standard here

tiborvass commented:
I think design-wise we agree so moving to code review.

cpuguy83 commented:
Any chance you can split the swarmkit vendor into a separate commit?

Commit:
Replace secrets with join tokens

Implement the proposal from
https://github.com/docker/docker/issues/24430#issuecomment-233100121

Removes acceptance policy and secret in favor of an automatically
generated join token that combines the secret, CA hash, and
manager/worker role into a single opaque string.

Adds a docker swarm join-token subcommand to inspect and rotate the
tokens.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
Split the vendoring commits and addressed some early feedback. Starting to work on updating tests.

diogomonica commented on api/client/swarm/join.go:
```diff
@@ -46,20 +42,16 @@ func runJoin(dockerCli *client.DockerCli, opts joinOptions) error {
-	} else {
-		fmt.Fprintln(dockerCli.Out(), "This node joined a Swarm as a worker.")
-	}
+	// TODO(aaronl): Perhaps SwarmJoin should return whether the node
```
Agreed.

diogomonica commented on api/client/swarm/join_token.go:
```diff
@@ -0,0 +1,71 @@
+				return err
+			}
+
+			if args[0] == "worker" {
```
This will allow rotating both tokens simultaneously. Do we want to have a different output that prefaces each token with `Worker:` and `Manager:` in the case both new tokens are being printed out?

aaronlehmann replied:
You can only rotate one at a time.

diogomonica commented:
LGTM

aaronlehmann commented:
Tests should be passing now. Also restored the `This node joined a Swarm as a (worker|manager)` message. Starting to work on doc changes.

aaronlehmann commented:
I've updated the API and CLI reference docs. @sfsmithcha @thaJeztah PTAL.

aaronlehmann commented:
engine-api PR is docker/engine-api#327.

swarmkit PR is docker/swarmkit#1189 (already merged). I'll fix the swarmkit vendoring once engine-api is ready to go as well.

Commit:
Revendor engine-api

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Revendor swarmkit

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
Both upstream PRs were merged. Updated the PR to properly vendor those components.

tonistiigi commented:
LGTM

sfsmithcha commented on docs/reference/commandline/node_ls.md:
```diff
@@ -54,16 +54,16 @@ The `name` filter matches on all or part of a node name.
     $ docker node ls -f id=1
-    ID                         HOSTNAME       MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS  LEADER
-    1bcef6utixb0l0ca7gxuivsj0  swarm-worker2  Accepted    Ready   Active
+    ID                         HOSTNAME       STATUS  AVAILABILITY  MANAGER STATUS  LEADER
```
Is this right? I don't see a `LEADER` column.

```
ID                           HOSTNAME      STATUS  AVAILABILITY  MANAGER STATUS
5ff7wbwlpntfir9o85ji525d2 *  2b2a7bb10c50  Ready   Active        Leader

aaronlehmann replied:
Thanks. Unrelated to my change, but I've corrected these examples.

sfsmithcha commented on docs/reference/commandline/swarm_init.md:
````diff
@@ -34,54 +32,22 @@ in the newly created one node Swarm cluster.
-```bash
-$ docker swarm init --listen-addr 192.168.99.121:2377 --auto-accept none
-```
+`docker swarm init` generates random tokens that can be used by other nodes to join the swarm as
````
This could be stronger:

`docker swarm init` generates two random tokens, a worker token and a manager token. When you join a new node to the swarm, the node joins as a worker node or manager node based upon the token you pass to `docker swarm join`.

After you create the swarm, you can display a token or rotate a token using [swarm join token](swarm_join_token.md).

sfsmithcha commented on docs/reference/commandline/swarm_join.md:
```diff
@@ -17,52 +17,52 @@ Usage:  docker swarm join [OPTIONS] HOST:PORT
 
-Join a node to a Swarm cluster. If the `--manager` flag is specified, the docker engine
-targeted by this command becomes a `manager`. If it is not specified, it becomes a `worker`.
+Join a node to a Swarm cluster. If the token given to the `--token` flag is a manager token,
```
Swarm is lowercase except at the beginning of a sentence.  Suggest:

Join a node to a swarm. The node joins as a manager node or worker node based upon the token you pass with the `--token` flag. If you pass a manager token, the node joins as a manager. If you pass a worker token, the node joins as a worker.

aaronlehmann replied:
I'm afraid `Swarm` isn't consistently lowercased. The "swarm join" reference page uses a mix of title case and lower case, and messages from the CLI itself tend to use title case; for example "This node joined a Swarm as a worker.".

Should I go through and change these?

tiborvass replied:
I suggest fixing casing in a separate docs PR to make it consistent everywhere.

aaronlehmann replied:
I'll just do it here if you don't mind. It's not many changes, but I'm worried about conflicts if I make a separate PR.

sfsmithcha replied:
@tiborvass , @aaronlehmann , either way, we'll need to review the doc set for consistent usage in addition to this PR. So I've opened a docs issue: https://github.com/docker/docker/issues/24905

sfsmithcha commented on docs/reference/commandline/swarm_join.md:
```diff
@@ -17,52 +17,52 @@ Usage:  docker swarm join [OPTIONS] HOST:PORT
 
 ### Join a node to swarm as a manager
 
+In the example below, the token given to `--token` is a manager token.
```
The example below demonstrates using a manager token.

sfsmithcha commented on docs/reference/commandline/swarm_join.md:
```diff
@@ -17,52 +17,52 @@ Usage:  docker swarm join [OPTIONS] HOST:PORT
+
 ### Join a node to swarm as a worker
 
+This command line is the same as the example above, except that the token is a worker token, which
```
The example below demonstrates joining a worker node using a worker token.

sfsmithcha commented on docs/reference/commandline/swarm_join_token.md:
````diff
@@ -0,0 +1,52 @@
+      --rotate   Rotate join token
+```
+
+Join tokens are the strings passed to the `--token` flag of `swarm join`. This command allows the
````
Join tokens are secrets that determine whether or not a node will join the swarm as a manager node or a worker node. You pass the token using the `--token` flag when you run `swarm join`. You can access the current tokens or rotate the tokens using `swarm join-token`.

sfsmithcha commented on docs/reference/commandline/swarm_join_token.md:
```diff
@@ -0,0 +1,52 @@
++++
+<![end-metadata]-->
+
+# swarm init
```
# swarm join-token

sfsmithcha commented on docs/reference/commandline/swarm_join_token.md:
````diff
@@ -0,0 +1,52 @@
+SWMTKN-1-3pu6hszjas19xyp7ghgosyx9k8atbfcr8p2is99znpy26u2lkl-7p73s1dx5in4tatdymyhg9hu2
+```
+
+When used with the `--rotate` flag, a new join token will be generated for the specified role,
````
Use the `--rotate` fag to generate a new join token for the specified role and print the output:

sfsmithcha commented on docs/reference/commandline/swarm_join_token.md:
```diff
@@ -0,0 +1,52 @@
+
+### `--rotate`
+
+This flag causes the join token for the specified role to be replaced with a new token.
```
Update the join token for a specified role with a new token and print the token.

sfsmithcha commented:
Aside from cp/paste error in header of `swarm join-token` and editorial suggestions LGTM

aaronlehmann commented:
Added a commit to address docs feedback. Will squash this once it has been reviewed.

I fixed title case references to `Swarm` in the actual CLI and the CLI reference docs. There are still quite a few in other docs. These can be addressed in a separate PR.

aaronlehmann commented:
I've updated `docker swarm join-token` to print a full `docker swarm join` command by default, per the suggestion in #24907. With the `-q` or `--quiet` option, it only prints the token.

PTAL.

tiborvass commented:
Small nit, but it feels weird to have `join` and `join-token` not next to each other in `docker swarm`'s output.

tiborvass commented:
Otherwise LGTM

SvenDowideit commented:
this appears to have been cherry-picked into bump_v1.12.0 already.
